weave/tests/test_cond.py

```diff
@@ -0,0 +1,18 @@
+import weave
+from weave import ops_arrow
+
+
+def test_cond_basic():
+    assert weave.use(weave.ops.cond({"a": True}, [5])) == 5
+    assert weave.use(weave.ops.cond({"a": False}, [5])) == None
+    assert weave.use(weave.ops.cond({"a": False, "b": True}, [5, 6])) == 6
+
+
+def test_cond_vector():
+    conds = weave.save(
+        ops_arrow.to_arrow(
+            [{"a": True, "b": False}, {"a": False, "b": False}, {"a": False, "b": True}]
+        ),
+        "conds",
+    )
+    assert weave.use(conds.cond([5, 6])).to_pylist_raw() == [5, None, 6]
```

Comment:
Could we consider an alternative approach to the API? I find it a bit challenging to understand, and I have a suggestion that might make it more intuitive. How about something like the following?: 

```
AWL([1, 2, 3]).if(cond=lambda val: val == 2, then=lambda val: val + 2, else=lambda val: val - 1)   -> [0, 4, 2]
```

While this approach isn't exactly the same (it is a ternary operator instead of a switch statement), I believe incorporating a ternary operator into the language (e.g., treesitter as ? :) could add value. It could even emulate a switch statement with nesting.

This could still be vectorized efficiently.

The algorithm here would be

1. Vectorize `lambda val: val == 2`
2. Apply it to the input list, returning AWL<bool>[False, True, False]
3. Split the AWL into two sub awls using `awl.filter()` ->` true_list = AWL<int>[2]`, `false_list = AWL<int>[1, 3]`.
4.  Vectorize the true and false functions, `lambda val: val + 2`, `lambda val: val - 1`
5. Apply them to `true_list` and `false_list`, respectively, giving [4], [0, 2]. 
6. Use `pc.replace_with_mask()` and the computed filter (and its inverse) to update the original array.

By doing this, we could maintain full vectorization without any unnecessary computation, and make the API a bit more user-friendly. Moreover, by avoiding case_when, it  will set us to better handle UnionArrays and other intricate types in the future. Also, it fully covers the  boolean space of possibilities, ensuring that every list element is explicitly handled in a branch.  What do you think?

Reply:
Agreed the cond variant is not intuitive to write. If you look at the PR again, you'll I added a helper function to make it nicer (take a look at the panel_openai_monitor cost function again).

Definitely like the thinking here, I had thought a bit about this too. We could even have compile detect a sequence of these and convert it to a case when if we want.

Using lambdas it a bit more complicated, we need to update stitch for example to traverse them. (lambdas generally make the graph harder to statically analyze)

I also think that counterintuitively, the proposal might be slower then the case_when version, even though the case_when one does significant amounts of over-calculation. Shuffling memory around (in the split step) is more expensive than just jamming a vector through a simd operation. The tradeoffs depend on the shape of the inputs though.

Reply:
I just added a fun little notebook to the PR to explore the perf tradeoffs here: https://github.com/wandb/weave/blob/5b8230c2e78c1348474a4d567b0ac43fbb86ff7c/perf_notebooks/case_when%20v%20if_then_else.ipynb

On my M1, the if_then_else construction often beats the case_when variant, and beats it by a lot (order of magnitude or more) when an early branch eliminates most of the input. Sometimes the case_when variant wins, often they are close (within 20%). This could change a lot depending cpu architecture.

The case_when flavor is much easier to parallelize though, you just compute all branches in parallel. 

I the think the user API with the if_then_else variant is gnarlier though because of the lambdas (although we could also prob wrap in helpers), and having more lambdas in in the graph is generally more annoying, so I think I'll leave the implementation I have. We can add an if_then_else later if it keeps coming up.

Reply:
Personally I feel like adding `lamdba : ` in python is worth the tradeoff, and then we could add a ternary operator to treesitter that makes it so you dont have to write lambdas in the UI. The better perf is also compelling, thanks for making the notebook, thats super cool. But approving to unblock as I know we are on a tight deadline to ship